runtime/compiler/control/JITaaSCompilationThread.cpp

```diff
@@ -822,43 +833,80 @@ bool handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe)
             }
-         // Cannot recreate thunk on the client because code cache is not yet allocated.
-         // Save it until compilation ends and relocateThunks is called
-         compInfoPT->addThunkToBeRelocated(serializedThunk, signature);
-         client->write(response, JITServer::Void());
+         else
+            {
+            // For AOT, set address to received string, because it will be stored to SCC, so
+            // no need for code cache allocation
+            thunkAddress = reinterpret_cast<void *>(&serializedThunk[0] + 8);
+            }
+
+         // Ideally, should use signature.data() here, but setJ2IThunk has non-const pointer
+         // as argument, and it uses it to invoke a VM function that also takes non-const pointer.
+         thunkAddress = fe->setJ2IThunk(&signature[0], signature.size(), thunkAddress, comp);
+
+         client->write(response, thunkAddress);
+         }
+         break;
+      case MessageType::VM_needsInvokeExactJ2IThunk:
+         {
+         auto recv = client->getRecvData<std::string>();
+         std::string signature = std::get<0>(recv);
+
+         TR_J2IThunkTable *thunkTable = comp->getPersistentInfo()->getInvokeExactJ2IThunkTable();
+         // Ideally, should use signature.data() here, but findThunk takes non-const pointer
+         TR_J2IThunk *thunk = thunkTable->findThunk(&signature[0], fe);
+         client->write(response, thunk == NULL);
          }
          break;
       case MessageType::VM_setInvokeExactJ2IThunk:
          {
          auto recv = client->getRecvData<std::string>();
          std::string &serializedThunk = std::get<0>(recv);
 
-         if (comp->compileRelocatableCode())
+         uint8_t *thunkStart;
+         if (!comp->compileRelocatableCode())
             {
-            // For AOT, store thunk to SCC immediately, in case relocations get delayed
-            static_cast<TR_J9SharedCacheVM *>(fe)->persistJ2IThunk(reinterpret_cast<void *>(&serializedThunk[0]));
+            // Same as regular thunks, copy thunk to code cache for regular compilation,
+            // save to SCC for AOT compilation.
+            // Do not need relocation here, because helper address should have been originally
+            // fetched from the client.
+            thunkStart = TR_JITaaSRelocationRuntime::copyDataToCodeCache(serializedThunk.data(), serializedThunk.size(), fe);
+            if (!thunkStart)
+               compInfoPT->getCompilation()->failCompilation<TR::CodeCacheError>("Failed to allocate space in the code cache");
             }
-         // Cannot recreate thunk on the client because code cache is not yet allocated.
-         // Save it until compilation ends and relocateThunks is called
-         TR::compInfoPT->addInvokeExactThunkToBeRelocated(serializedThunk);
-         client->write(response, JITServer::Void());
+         else
+            {
+            thunkStart = reinterpret_cast<uint8_t *>(&serializedThunk[0]);
```

Comment:
I remember that invokeExact thunks are always allocated from code cache, so we must copyDataToCodeCache in this case as well. Using the address from the serialized thunk will not work.

Reply:
I guess we should copy it to code cache for consistency, but in AOT, this thunk will be immediately stored to SCC, so it doesn't matter where it's originally allocated.

Reply:
> but in AOT, this thunk will be immediately stored to SCC

MethodHandles aren't supported in AOT.

Reply:
Yeah, but why do we still have relocations and SCC stores for invoke exact thunks?

Reply:
> Yeah, but why do we still have relocations and SCC stores for invoke exact thunks?

I have no idea, but it would never work as the codebase stands; Method Handle support has yet to be added. Actually if you look for `TR_J2IThunks` in `RelocationRecord.cpp` there is no associated ReloRecord for it. Grepping for it in the codebase shows some places adding an external relocation, but those code paths are never executed in AOT lol. 

As for `TR_Thunks` and `TR_J2IVirtualThunkPointer`, they're used to relocate a Snippet which has a pointer to a J2I thunk; which is different from an Invoke* thunk (which has to do with Method Handles).

Reply:
So do you think we should keep doing whatever mainline does, assuming MethodHandle support will be eventually enabled for AOT, or just ignore it for now?

Reply:
> So do you think we should keep doing whatever mainline does

Yes, however, when the compiler on the server goes to register the thunks, you're going to need to:

1. send the thunk over to the client
2. have the client allocate some code cache memory and memcopy the thunk
3. register the thunk
4. send the thunk pointer (which is valid in the client's address space) over to the server

I think you have to do this for J2I and InvokeExactJ2I thunks.

Reply:
We already do that for non-AOT, excluding sending client's invokeExactJ2IThunk pointer back to the server, because the server will never use it anyway. 
But mainline AOT does not register the thunk, it only saves it to SCC, so wouldn't we want to keep behaviour identical?

Reply:
> But mainline AOT does not register the thunk, it only saves it to SCC, so wouldn't we want to keep behaviour identical?

Yeah it does, assuming that when server stores to the SCC it results in it getting stored to the client's SCC.

Reply:
Yes, it's stored to client's SCC.